### ISSUE TITLE
SDK Server: Adopt Sidecar Containers

### ISSUE THREAD

**Issue by markmandel** [googleforgames/agones#3642]
**Is your feature request related to a problem? Please describe.**

The current [Health Failure Strategy and lifecycle](https://agones.dev/site/docs/guides/health-checking/#health-failure-strategy) for the sdkserver and game server container works, but it has edge cases, and is eventually consistent - which can also be fun (but not always in a good way).

This means as new ways for Pods to fail come up we have build out new features to capture this, and it's not straightforward, and often puts more load on the K8s control plane.

**Describe the solution you'd like**

Once Agones supports Kubernetes 1.29+ (or maybe earlier behind a feature flag?) we could move the sdkserver to the new [Sidecar container](https://kubernetes.io/docs/concepts/workloads/pods/sidecar-containers/) model.

Then we can set the Pod to `restart: Never` by default, and the sdkserver sidecar to `restart: Always`, which would simplify things greatly.

* Alot of https://github.com/googleforgames/agones/blob/main/pkg/gameservers/health.go would likely dissapear, especially around setting `GameServerReadyContainerIDAnnotation` to know if the container is ready or not.
* The sdkserver only shutting down on `Shutdown` state also goes away, since the sidecar container can be spun up before and stay alive for the entire duration of the Pod.

There is one thing that is definitely tricky here - the way we have things setup now, if a GameServer is before Ready, we let it restart. If it's after Ready, we do not. Maybe we should revisit this as a pattern we still need for simplification - especially if the Pod crashes before being Ready, a new one will be recreated 🤔 Although this breaks backward compatibility, and that is less than ideal.

**Describe alternatives you've considered**
Leave things as they are, they seem to mostly work!

**Additional context**

N/A

**Comment by markmandel** [comment]
Just copying a comment from an offline conversation with a user -- "the atomic non-restart to Unhealthy is way more important to us than being able to restart a game server process before it moves to Ready".

**Comment by swermin** [comment]
We encountered this problem recently and I do agree that simplifying this flow would be a tremendous help.
We have our pods set to `restart: Never` since we do want the game servers to cycle as fast as possible. But with this setup if the game server `terminates` before reaching the `Ready` state then that pod is forever in a state where we cannot re-use it.
I do understand the necessecity of being able to restart a pod if it terminates before reaching the ready state, but in our case the only thing that can make the server terminate is if we have a crash so a restart always would just make the whole pod be in crashloop. To solve this quirk we need some hacks with `prestop` hooks to get it to be recycled properly.

Long story short, this is a good change!

**Comment by aRestless** [comment]
With Agones 1.45.0 having bumped the minimum Kubernetes version to v1.29 this should now be safe to introduce.

As it currently stands, a crashing game server container (or one that simply didn't call agones.Shutdown()) will be restarted and then torn down shortly after, which has created quite some noise in operations for our projects in the past.

So this change would certainly help us and allow us to remove some "restart-in-place" protection workarounds in our code.

**Comment by towolf** [comment]
What's the issue with `restart: Never`? If we always reap Pods instead of having them restart, what's the drawback? I mean other than that it's a change of behaviour?

**Comment by markmandel** [comment]
> What's the issue with restart: Never? If we always reap Pods instead of having them restart, what's the drawback? I mean other than that it's a change of behaviour?

If the sidecar SDK server crashes for some reason, your SDK just stops working. Ideally it would restart and be self healing.

**Comment by markmandel** [comment]
Yep - that's the proposed change 👍🏻

**Comment by markmandel** [comment]
Started working on this a bit. Seemed like a fun time 😄

**Comment by markmandel** [comment]
Just letting people know I'm still working on this 👍🏻 made some more progress yesterday.

**Comment by markmandel** [comment]
Getting there! Got a few end to end years to fix. But the net positive will be that once this becomes stable we can delete a good chunk of code 👍🏻

**Comment by swermin** [comment]
That’s great to hear! Can’t wait 🙌

**Comment by markmandel** [comment]
There we go! That PR should be the working implementation. Once that's done, I can get the docs updated and round this out 👍🏻
